src/python/pants/engine/legacy/parser.py

```diff
@@ -125,5 +130,35 @@ def parse(self, filepath, filecontent):
     # _intentional_ mutation would require a deep clone, which doesn't seem worth the cost at
     # this juncture.
     self._parse_context._storage.clear(os.path.dirname(filepath))
-    six.exec_(python, dict(self._symbols))
+
+    symbols = dict(self._symbols)
+    if self._build_file_import_behavior != 'allow':
```

Comment:
It seems like this whole block should be in `def _generate_symbols` in order to avoid running it for every parse?

You won't be able to get access to the complete filepath, but you should be able to lazily access `ParseContext.rel_path`, which is mutated for each parse.

Reply:
Done. `ParseContext.rel_path` unfortunately is just the directory, not the file, which makes the warning slightly less useful... Added "in directory" to the warning message.

Reply:
You could add another (private) field to the `ParseContext` if it helps.